rotkehlchen/chain/ethereum/modules/aave/v3/balances.py

```diff
@@ -0,0 +1,129 @@
+from rotkehlchen.constants.resolver import evm_address_to_identifier
+from rotkehlchen.errors.asset import UnknownAsset, WrongAssetType
+from rotkehlchen.errors.misc import RemoteError
+from rotkehlchen.history.events.structures.types import HistoryEventSubType, HistoryEventType
+from rotkehlchen.inquirer import Inquirer
+from rotkehlchen.logging import RotkehlchenLogsAdapter
+from rotkehlchen.types import ChainID, EvmTokenKind, SupportedBlockchain
+
+if TYPE_CHECKING:
+    from rotkehlchen.chain.ethereum.interfaces.balances import BalancesType
+    from rotkehlchen.chain.ethereum.node_inquirer import EthereumInquirer
+    from rotkehlchen.db.dbhandler import DBHandler
+    from rotkehlchen.types import ChecksumEvmAddress
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+
+class Aavev3Balances(ProtocolWithBalance):
+    def __init__(self, database: 'DBHandler', evm_inquirer: 'EthereumInquirer'):
+        super().__init__(
+            database=database,
+            evm_inquirer=evm_inquirer,
+            counterparty=CPT_AAVE_V3,
+            deposit_event_types={(HistoryEventType.DEPOSIT, HistoryEventSubType.DEPOSIT_ASSET)},
+        )
+        self.balances = BlockchainBalances(db=database).get(SupportedBlockchain.ETHEREUM)
+
+    def query_balances(self) -> 'BalancesType':
+        """
+        Query balances for Aave v3 open positions.
+        Docs: https://docs.aave.com/developers/guides/rates-guide#how-is-yield-accrued
+
+        The positions are fetched from the AToken contracts by calling the .balanceOf()
+        method on them. Then return the remaining balances that is not detected.
+        """
+        balances: BalancesType = defaultdict(lambda: defaultdict(Balance))
+        addresses_events = self.addresses_with_activity(
+            products=None,
+            event_types=self.deposit_event_types,
+        )
+        if len(addresses_events) == 0:
+            return balances
+
+        users_tokens: dict['ChecksumEvmAddress', set['ChecksumEvmAddress']] = defaultdict(set)  # unique tokens for each address  # noqa: E501
+        for user_address, events in addresses_events.items():
+            for event in events:
+                if event.address is not None:
+                    users_tokens[user_address].add(event.address)
+
+        # prepare the arguments for the contract calls and the encoded call
+        calls: list[tuple['ChecksumEvmAddress', str]] = []
+        calls_arguments: list['ChecksumEvmAddress'] = []
+        address_to_contract: dict['ChecksumEvmAddress', 'EvmContract'] = {}
+        for user_address, token_addresses in users_tokens.items():
+            for token_address in token_addresses:
+                address_to_contract[token_address] = EvmContract(
+                    address=token_address,
+                    abi=self.evm_inquirer.contracts.erc20_abi,
+                    deployed_block=0,  # not used
+                )
```

Comment:
you only use the contract to encode and afaik the address doesn't matter, only the abi. Please double check it but I believe you only need to create once the EvmContract and use it to encode all the calls. I kind of remember that we do this somewhere